### ISSUE TITLE
API to read an EC public key from momory buffer (DER format)

### ISSUE THREAD

**Issue by mishraab07** [ARMmbed/mbedtls#2560]
**Question:** I am having difficulties in finding an API that can read an EC public key (and not X509 public certificate) from memory buffer (DER format) and gives back the mbed internal ecdsa context (for ex: mbedtls_ecdsa_context).

The requirement is that I have is a ECDSA signature and memory buffer containing EC public key in DER format, and the task is to verify the ECDSA signature using the given public key.

In short, I will be requiring an API in mbedTLS, which should-be/is very similar to openSSL's d2i_EC_PUBKEY(). 

Any leads would be very helpful, thank you.

**Comment by RonEld** [comment]
Hi @mishraab07 
Thank you for your question.
I believe that [this example](https://github.com/ARMmbed/mbedtls/blob/development/programs/pkey/key_app.c) should show you how to parse a public key.
It will also show you [how to get the `mbedtls_ecp_keypair` struct](https://github.com/ARMmbed/mbedtls/blob/development/programs/pkey/key_app.c#L289)

**Comment by mishraab07** [comment]
Thanks for the pointer, I am unblocked.

I do have one more question:  Can we get public key from a given mbedtls_ecdsa_context; I know there are APIs available that takes mbedtls_pk_context as  input parameter & returns  der encoded data. But I am not able to find an API that transforms mbedtls_ecdsa_context to mbedtls_pk_context.  

Thank you again

**Comment by RonEld** [comment]
@mishraab07 
unfortunately, there isn't such an APi.
However, could [this post](https://forums.mbed.com/t/write-rsa-public-key-top-pem/4524/2) answer your question?

**Comment by mishraab07** [comment]
Thanks for the link @RonEld .

However, I am not very clear about using **mbedtls_ecp_point_write_binary()** in my case.
Does mbedtls_ecp_point_write_binary() fill the buffer supplied by caller with DER encoded public key ?

My requirement is exactly opposite of what mbedtls_pk_ec() does i.e. given an mbedtls_ecdsa_context, I need to write the EC public key (DER format) associated with the mbedtls_ecdsa_context into a buffer supplied ?

**Comment by RonEld** [comment]
As you mentioned, in order to write the EC public key, you will need to use `mbedtls_pk_write_pubkey_der`, which receives `mbedtls_pk_context` as parameter.
You will need to set the `mbedtls_pk_ec()` of the pk context to have the parameters of your `mbedtls_ecdsa_context`.
You don't need to use `mbedtls_ecp_point_write_binary` as you don't have \ need to keys in binary format, however in the linnk i showed you there is also explanation on how to use the rsa context into the pk context. Doing the similar thing with ecdsa context is same principle.
You will need to do something of the following ( this is an untested example):
~~~
mbedtls_pk_context pk;

mbedtls_pk_setup( &pk, mbedtls_pk_info_from_type(MBEDTLS_PK_ECDSA) );
mbedtls_ecp_copy( &mbedtls_pk_ec(pk).Q, &ecdsa_ctx.Q);
mbedtls_mpi_copy(&mbedtls_pk_ec(pk).d, &ecdsa_ctx.d);
mbedtls_pk_ec(pk).grp =ecdsa_ctx.grp;
~~~

**Comment by mishraab07** [comment]
Thanks for this, it was very useful, however in this implementation, I am failing at **mbedtls_pk_write_pubkey_der()** function call. 

My code snippet looks like this:

mbedtls_pk_context pk_ctx;
pk_ctx.pk_info = mbedtls_pk_info_from_type(MBEDTLS_PK_NONE);

result = mbedtls_pk_setup(&pk_ctx, mbedtls_pk_info_from_type(MBEDTLS_PK_ECDSA));
TEST_ASSERT_EQUAL(0, result);

mbedtls_ecp_copy(&(mbedtls_pk_ec(pk_ctx)->Q), &(avalidkey.Q));
mbedtls_mpi_copy(&(mbedtls_pk_ec(pk_ctx)->d), &(avalidkey.d));
mbedtls_pk_ec(pk_ctx)->grp = avalidkey.grp;

key_buf_len = mbedtls_pk_write_pubkey_der(&pk_ctx, p, 512);

if(key_buf_len <= 0){
    char err[100] = {0};
    char *p_e = err;
    mbedtls_strerror(key_buf_len, p_e, 100);
    printf("Error code: %s\n", p_e);
    result = 1;
    TEST_ASSERT_EQUAL(0, 1);
}

The erro code didn't helped much either i.e. **Error code: PK - Unavailable feature, e.g. RSA disabled for RSA key**

Does this mean, we can't proceed further ?

**Comment by mishraab07** [comment]
I figured it out as soon as I wrote back to you.
The problem was while setting up pk context.

Changing line:
result = mbedtls_pk_setup(&pk_ctx, mbedtls_pk_info_from_type(MBEDTLS_PK_ECDSA));
to
result = mbedtls_pk_setup(&pk_ctx, mbedtls_pk_info_from_type(MBEDTLS_PK_ECKEY));

worked as expected.

Thanks a lot for your time @RonEld , Have a nice day.

**Comment by RonEld** [comment]
@mishraab07 I am glad it worked for you!

I am closing as this question was answered and resolved. If you think this issue was closed by mistake, please reopen.
